contracts/Polygon_SpokePool.sol

```diff
@@ -0,0 +1,81 @@
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "./SpokePool.sol";
+import "./SpokePoolInterface.sol";
+import "./PolygonTokenBridger.sol";
+
+// ERC20s (on polygon) compatible with polygon's bridge have a withdraw method.
+interface PolygonIERC20 is IERC20 {
+    function withdraw(uint256 amount) external;
+}
+
+// IFxMessageProcessor represents interface to process messages.
+interface IFxMessageProcessor {
+    function processMessageFromRoot(
+        uint256 stateId,
+        address rootMessageSender,
+        bytes calldata data
+    ) external;
+}
+
+/**
+ * @notice Polygon specific SpokePool.
+ */
+contract Polygon_SpokePool is SpokePoolInterface, IFxMessageProcessor, SpokePool {
+    using SafeERC20 for PolygonIERC20;
+    address public fxChild;
+    PolygonTokenBridger public polygonTokenBridger;
+    bool private callValidated = false;
+
+    event PolygonTokensBridged(address indexed token, address indexed receiver, uint256 amount);
+
+    constructor(
+        PolygonTokenBridger _polygonTokenBridger,
+        address _crossDomainAdmin,
+        address _hubPool,
+        address timerAddress
+    ) SpokePool(_crossDomainAdmin, _hubPool, 0x4200000000000000000000000000000000000006, timerAddress) {
+        polygonTokenBridger = _polygonTokenBridger;
+    }
+
+    function processMessageFromRoot(
+        uint256,
+        address rootMessageSender,
+        bytes calldata data
+    ) public {
+        // Validation logic.
+        require(msg.sender == fxChild, "Not from fxChild");
+        require(rootMessageSender == hubPool, "Not from mainnet HubPool");
+
+        // This sets a variable indicating that we're now inside a validated call.
+        // Note: this is used by other methods to ensure that this call has been validated by this method and is not
+        // spoofed.
+        callValidated = true;
+
+        // This uses delegatecall to take the information in the message and process it as a function call on this contract.
+        (bool success, ) = address(this).delegatecall(data);
+        require(success, "delegatecall failed");
+
+        // Reset callValidated to false to disallow admin calls after this method exits.
+        callValidated = false;
+    }
+
+    /**************************************
+     *        INTERNAL FUNCTIONS          *
+     **************************************/
+
+    function _bridgeTokensToHubPool(RelayerRefundLeaf memory relayerRefundLeaf) internal override {
+        PolygonIERC20(relayerRefundLeaf.l2TokenAddress).safeIncreaseAllowance(relayerRefundLeaf.amountToReturn);
+        polygonTokenBridger.send(relayerRefundLeaf.l2TokenAddress, relayerRefundLeaf.amountToReturn);
+
+        emit PolygonTokensBridged(relayerRefundLeaf.l2TokenAddress, address(this), relayerRefundLeaf.amountToReturn);
+    }
+
+    function _requireAdminSender() internal view override {
+        require(callValidated, "Must call processMessageFromRoot");
```

Comment:
This is a clever way to check that calls are only made via the fx tunnel. But could you not replace this call with a require that the msg.sender is the fxChild and remove the callValidated Boolean variable?

Reply:
Yes! This was something that I considered. The particular case that I'm worried about (and maybe I shouldn't be) is that the polygon fxChild contract could make _other_ calls into contracts (now or in the future) on behalf of mainnet callers. If 1) those methods happen to be the same name/args as one of our admin methods or 2) (more likely) there's a way to partially or fully customize the selector used (hash of method name + arg types), someone could make a malicious call that comes from that contract but hasn't had its sender validated.

The types of attacks I'm worried about are ones in the style of the poly network hack, where someone is able to use user-inputs to force a contract to call another contract in a way that it wasn't intended to. https://twitter.com/kelvinfichter/status/1425217065690992641?s=20&t=OL5ecSja3ii-LktomykOOg.

I doubt this is possible on polygon today, but their contracts are upgradable and I just figure it's better to be extra careful.

Does that make sense? Thoughts?

Reply:
I think that's a reasonable concern and one worth documenting! Two comments:
- I think you should add the `callValidated` flipping on/ff as a modifier for readability
- You should add a comment explaining why checking msg.sender == fxChild is not sufficient because of the reasons you explained above

Reply:
Done.